### PR TITLE
[dist] fix: the issue where NPU does not support the PreSumMul operation

### PR DESCRIPTION
**Pull request by Feng0w0**
### What does this PR do?

NPU does not support ReduceOp.PREMUL_SUM, so a patch is used to split the operator into ReduceOp.SUM and a multiplication operation

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `misc`, `ci`, `config`, `docs`, `core`, `data`, `dist`, `omni`, `logging`, `model`, `optim`, `ckpt`, `release`, `task`, `
  - If this PR involves multiple modules, separate them with `,` like `[megatron, torch, liger-kernals,fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, torch, liger-kernals] feat: dynamic batching`

### Test

I used Qwen3VL-MOE to test the accuracy and performance of my method on  8 A100.
#### 1.precision
Activate deterministic and use default method
```Plain Text
Epoch 1/2:   1%|          | 5/500 [07:22<9:46:12, 71.05s/it, loss: 0.4793784022, grad_norm: 6.9401021004, lr: 1.00e-05] Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   1%|          | 6/500 [07:48<7:38:04, 55.64s/it, loss: 0.3809688091, grad_norm: 15.4080181122, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   1%|▏         | 7/500 [08:14<6:19:05, 46.14s/it, loss: 0.3065966070, grad_norm: 5.9091300964, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 8/500 [08:40<5:24:27, 39.57s/it, loss: 0.1315839440, grad_norm: 6.1710629463, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 9/500 [09:07<4:51:31, 35.62s/it, loss: 0.0580307879, grad_norm: 3.0772423744, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 10/500 [09:33<4:25:53, 32.56s/it, loss: 0.0247170310, grad_norm: 1.7002766132, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 11/500 [09:59<4:10:48, 30.77s/it, loss: 0.0104421172, grad_norm: 0.9332838655, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 12/500 [10:25<3:58:32, 29.33s/it, loss: 0.0056380886, grad_norm: 0.7326376438, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   3%|▎         | 13/500 [10:52<3:52:12, 28.61s/it, loss: 0.0048087277, grad_norm: 0.8172971010, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   3%|▎         | 14/500 [11:18<3:44:47, 27.75s/it, loss: 0.0026786863, grad_norm: 0.3342948556, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   3%|▎         | 15/500 [11:44<3:40:32, 27.28s/it, loss: 0.0022356007, grad_norm: 1.2753716707, lr: 9.99e-06]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
```
Activate deterministic and use my method：
```Plain Text
Epoch 1/2:   1%|          | 5/500 [07:16<9:25:44, 68.57s/it, loss: 0.4793784022, grad_norm: 6.9401021004, lr: 1.00e-05] Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   1%|          | 6/500 [07:42<7:24:30, 53.99s/it, loss: 0.3809688091, grad_norm: 15.4080181122, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   1%|▏         | 7/500 [08:09<6:10:12, 45.06s/it, loss: 0.3065966070, grad_norm: 5.9091300964, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 8/500 [08:34<5:18:36, 38.85s/it, loss: 0.1315839440, grad_norm: 6.1710629463, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 9/500 [09:01<4:46:45, 35.04s/it, loss: 0.0580307879, grad_norm: 3.0772423744, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 10/500 [09:27<4:22:58, 32.20s/it, loss: 0.0247170310, grad_norm: 1.7002766132, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 11/500 [09:54<4:09:01, 30.55s/it, loss: 0.0104421172, grad_norm: 0.9332838655, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   2%|▏         | 12/500 [10:19<3:56:56, 29.13s/it, loss: 0.0056380886, grad_norm: 0.7326376438, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   3%|▎         | 13/500 [10:47<3:51:50, 28.56s/it, loss: 0.0048087277, grad_norm: 0.8172971010, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   3%|▎         | 14/500 [11:12<3:44:21, 27.70s/it, loss: 0.0026786863, grad_norm: 0.3342948556, lr: 1.00e-05]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Epoch 1/2:   3%|▎         | 15/500 [11:39<3:40:28, 27.27s/it, loss: 0.0022356007, grad_norm: 1.2753716707, lr: 9.99e-06]Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
Too many dataloader workers: 2 (max is dataset.n_shards=1). Stopping 1 dataloader workers.
```
#### 2.performance
Each set of configurations will run three experiments, taking 5-15 steps of time.
|                                          | use my method |not use |
| --------------------------- | ----------- |----------- |
| activate deterministic      | 4m23, 4m23, 4m22 |4m22, 4m22, 4m23|
| inactivate deterministic   | 2m46, 2m43, 2m42 |2m44, 2m44, 2m43  |

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `make commit && make style && make quality`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by Feng0w0** [comment]
add ascend lable

**Comment by Feng0w0** [comment]
This PR inherits from:https://github.com/ByteDance-Seed/VeOmni/pull/245

File: veomni/distributed/torch_parallelize.py [hunk 1]
```diff
@@ -349,16 +354,13 @@ def _experts_shard_placement_fn(param):
...
-            # NPU is missing PreSumMul ReduceOp for `set_gradient_divide_factor` API
-            # As a result, we will handle the grad dividing inisde the grad norm clipping
-            # Need to remove this condition after the issue is resolved.
-            if not IS_NPU_AVAILABLE:
-                gradient_divide_factor = parallel_state.ep_gradient_divide_factor
-                logger.info(f"setting grad divide factor for ep module to {gradient_divide_factor}")
+            # but for torch 2.7 we still use set_reduce_scatter_divide_factor(parallel_state.ep_size)
+            gradient_divide_factor = parallel_state.ep_gradient_divide_factor
+            logger.info(f"setting grad divide factor for ep module to {gradient_divide_factor}")
+            if hasattr(experts_mod, "set_gradient_divide_factor"):
```
Comment (Luosuu) [inline veomni/distributed/torch_parallelize.py]: IMO, maybe we should use 
```python
if IS_NPU:
  # NPU is using torch 2.7
  experts_mod.set_reduce_scatter_divide_factor(gradient_divide_factor)
else:
 # from torch 2.8
  experts_mod.set_gradient_divide_factor(gradient_divide_factor)
```

Comment (Luosuu) [reply veomni/distributed/torch_parallelize.py]: We are forcing torch 2.8 on GPU anyway

File: veomni/distributed/torch_parallelize.py [hunk 2]
```diff
@@ -339,6 +339,11 @@ def _experts_shard_placement_fn(param):
     # | -- experts layer (apply fully_shard separately in order to shard across EP groups on the same EP rank instead of sharding globally)
     # | -- layers (declared in model.modules_to_ignore_in_mixed_precision) that need to apply fully_shard separately due to different mp policy as the decoder layer
     #      (e.g., some models requires MoE TopK gate layer to have parameters in higher FP32 precision in forward).
+    # NPU currently does not support the PreSumMul operation, so this operation is supported through the apply_hccl_premul_sum_patch.
```
Comment (piyifan123) [inline veomni/distributed/torch_parallelize.py:342]: nit: would you mind opening a github issue to clean this up once the HCCL side support is there? and add a TODO here to link to the github issue.

File: veomni/ops/patch/hccl_premul_sum.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import torch
+from torch.distributed.distributed_c10d import ReduceOp
+
+
+def hccl_premul_sum_wrapper(op, output_name):
```
Comment (piyifan123) [inline veomni/ops/patch/hccl_premul_sum.py:5]: nit: add docstring on what it does.

File: veomni/ops/patch/hccl_premul_sum.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        factor = None
+        if kwargs["op"] == ReduceOp.PREMUL_SUM:
+            factor = kwargs["op"].__getstate__()[1]
+            kwargs["op"] = ReduceOp.SUM
+        handle = op(*args, **kwargs)
+        if handle is not None:
+            handle.wait()
+        if factor is not None:
+            output = args[0] if len(args) > 0 else kwargs[output_name]
+            output.data.mul_(factor)
```
Comment (piyifan123) [inline veomni/ops/patch/hccl_premul_sum.py:24]: is the multiplication applied after the communication? if so, i guess the semantic is not the same (my understanding is that the behavior is multiply the factor and then communicate)? please add a comment to highlight this.

Comment (Feng0w0) [reply veomni/ops/patch/hccl_premul_sum.py:24]: Thank you for your suggestion.I have added.

File: veomni/ops/patch/hccl_premul_sum.py (hunk 3 above)
Comment (Feng0w0) [reply veomni/ops/patch/hccl_premul_sum.py:5]: Thank you for your suggestion.I have added.

File: veomni/distributed/torch_parallelize.py (hunk 2 above)
Comment (Feng0w0) [reply veomni/distributed/torch_parallelize.py:342]: Thank you for your suggestion.I have made the changes.I found that issue 241(https://github.com/ByteDance-Seed/VeOmni/issues/241) has already raised this issue, and I have added the link here.

File: veomni/distributed/torch_parallelize.py (hunk 1 above)
Comment (Feng0w0) [reply veomni/distributed/torch_parallelize.py]: Thank you for your suggestion.I have made the changes.
